app/controllers/concerns/abstract_xikolo_connector.rb

```diff
@@ -49,4 +50,60 @@ def handle_enrollments_response(response_data, user)
     end
     evaluate_enrollments_update_map update_map, user
   end
+
+  def get_dates_for_user(user)
+    token_string = "Legacy-Token token=#{get_access_token user}"
+    api_url = self.class::ROOT_API_V2 + DATES_API
+    response = RestClient.get(api_url, accept: 'application/vnd.xikoloapplication/vnd.xikolo.v1, application/json', authorization: token_string)
+    JSON.parse response
+  end
+
+  def handle_dates_response(response_data, user)
+    update_map = create_update_map_for_user_dates user, mooc_provider
+    response_data['dates'].each do |response_user_date|
+      user_date = UserDate.find_by(user: user, mooc_provider: mooc_provider, ressource_id_from_provider: response_user_date['resource_id'], kind: response_user_date['kind'])
+      if user_date.present?
+        update_map[user_date.id] = true
+        update_existing_entry user_date, response_user_date
+      else
+        create_new_entry user, response_user_date
+      end
+    end
+    change_existing_no_longer_relevant_entries update_map
+  end
+
+  def create_new_entry(user, response_user_date)
+    user_date = UserDate.new
+    user_date.user = user
+    user_date.mooc_provider = mooc_provider
+    user_date.course = Course.find_by(provider_course_id: response_user_date['course_id'])
```

Comment:
You're right, it is not a big overhead when it comes to the data stored in addition. Nevertheless, I still don't like it because I feel that it might lead to data inconsistency (if the mood provider here is different than it is in the course - which is very unlikely for the current implementation, but theoretical, our database supports it). 

In addition, I suggest to change one line in the method above: What about using:
``` ruby
course = Course.get_course_by_mooc_provider_id_and_provider_course_id(mooc_provider.id, response_user_date['course_id'])
user_date = UserDate.find_by(user: user, course: course, ressource_id_from_provider: response_user_date['resource_id'], kind: response_user_date['kind'])
```

This removes the usage of the mooc provider and, as far as I understand your implementation and your concerns, it should keep the assumption. In addition, it makes the resource_id_from_provider unique in the context of a course (instead of the whole provider). Of course, with the current implementation and the data we get from the Xikolo platforms, this should not be needed, but as we rely on external data, it feels better for me.